AmbientLightSensor: Update tests to be compliance with latest ED

Honry opened the pull request:
- Add https postfix as all interfaces must only be available within a secure context
 - Add new tests for onerror event, errored state, sensor object contruction, sensor readings
 - Update browsing context test
 - Reorg some tests

tobie commented on ambient-light/AmbientLightSensor_browsing_context.https.html:
```diff
@@ -6,26 +6,27 @@
+  sensor.onchange = t.step_func_done(function() {
+    assert_not_equals(sensor.reading, null);
+    let win = window.open('', '_blank');
+    assert_equals(sensor.reading, null);
```
This might probably change to hold on to the latest reading.

tobie commented on ambient-light/AmbientLightSensor_onchange.https.html:
```diff
@@ -5,26 +5,16 @@
-  sensor.onchange = t.step_func_done(function(event) {
+  sensor.onchange = t.step_func_done(function() {
+    assert_greater_than_equal(sensor.reading.illuminance, 0);
+    assert_equals(sensor.state, "active");
```
The state should now be called "activated".

tobie replied:
https://w3c.github.io/sensors/#dom-sensorstate-activated

tobie commented on ambient-light/AmbientLightSensor_onerror-manual.https.html:
```diff
@@ -0,0 +1,42 @@
+});
+
+async_test(function(t) {
+  sensor1.onstatechange = t.step_func_done(function(event) {
```
We're removing this from the spec.

tobie replied:
https://w3c.github.io/sensors/#dom-sensor-onactivate

tobie commented on ambient-light/AmbientLightSensor_reading.https.html:
```diff
@@ -0,0 +1,50 @@
+  let cachedReading = sensor.reading;
+  let cachedIlluminance = cachedReading.illuminance;
+  sensor.stop();
+  assert_equals(cachedReading.illuminance, cachedIlluminance);
```
You probably want to test for this asynchronously.

tobie commented on ambient-light/AmbientLightSensor_reading.https.html:
```diff
@@ -0,0 +1,50 @@
+}, "Test that sensor reading must be immutable.");
+
+test(function() {
+  assert_equals(cachedReading1, cachedReading2);
```
This should really be: `assert(cachedReading1 === cachedReading2);` or something similar, imho.

tobie commented on ambient-light/AmbientLightSensor_reading.https.html:
```diff
@@ -0,0 +1,50 @@
+  sensor1.stop();
+  assert_equals(sensor1.reading, null);
+  assert_equals(String(sensor2.reading), "[object AmbientLightSensorReading]");
+}, "Test that sensor is independent.");
```
I don't understand this test.

tobie commented on ambient-light/AmbientLightSensor_reading.https.html:
```diff
@@ -0,0 +1,50 @@
+    assert_not_equals(cachedReading2, cachedReading3);
+    sensor2.stop();
+  }, 2000);
+}, "Test that the sensor reading is updated when time passes.");
```
This is flaky. You should instead rely on a new `onchange` event.

tobie commented on ambient-light/AmbientLightSensor_start.https.html:
```diff
@@ -7,14 +7,32 @@
+}, "The default sensor.state is 'idle'");
+
+async_test(function(t) {
+  sensor.onstatechange = t.step_func_done(function(event) {
```
`onstatechange` has been removed.

tobie replied:
https://w3c.github.io/sensors/#dom-sensor-onactivate

Honry commented:
@tobie, many thanks for your comments, at present tests are compliance to latest [WD](https://www.w3.org/TR/generic-sensor/) in order to match what Chromium implementation follows. And I will update them once new WD published.

Honry committed:
AmbientLightSensor: Update tests to be compliance with latest WD
 - Add https postfix as all interfaces must only be available within a secure context
 - Add new tests for onerror event, errored state, sensor object contruction, sensor readings
 - Update browsing context test
 - Reorg some tests

tobie commented on ambient-light/AmbientLightSensor_stop.https.html:
```diff
@@ -8,15 +8,25 @@
 });
 
+async_test(function(t) {
+  sensor.onstatechange = t.step_func_done(function(event) {
```
https://w3c.github.io/sensors/#dom-sensor-onactivate

tobie commented:
@Honry please track the ED not the WD.

Honry commented:
@tobie, OK, I will update the tests later on.

Honry commented on ambient-light/AmbientLightSensor_stop.https.html:
```diff
@@ -8,15 +8,25 @@
+
 test(function() {
-  assert_equals(String(sensor.reading), "null");
+  assert_equals(sensor.reading, null);
```
@tobie, I would like to leave it as is until we make the final conclusion.

Honry commented:
@tobie, I have one concern about the condition to call `onchange` event.
Spec says, `onchange` event is called whenever a new reading is available. Meanwhile, according to the code snippet in your comment in [issues#15](https://github.com/w3c/ambient-light/issues/15#issuecomment-257409944), as my understanding, the sensor reading will change while time passing. But when I testing on Chromium, the `onchange` only can be called when the ambient light illuminance changes. Who is correct?

tobie commented:
This is still in flux.

Honry commented:
I think I should prioritize following the spec.  :)

Honry commented:
@tobie, Just update tests to latest ED, pls. help review. Thanks!

tobie commented on ambient-light/AmbientLightSensor_onerror-manual.https.html:
```diff
@@ -0,0 +1,32 @@
+});
+
+async_test(function(t) {
+  sensor.onactivate = t.step_func_done(function(event) {
```
Can't you just rewrite this as:

```js
 sensor.onactivate = t.step_func_done(assert_unreached);
```

Honry replied:
Of course, updated.

Honry committed:
Fix trailing whitespace

Honry committed:
Address @tobie's comments
 - Move reading stuff into onchange event
 - reading would hold on to the latest reading when the tab is inactive

Honry committed:
Update tests to latest ED
 - Remove onstatechange event tests
 - Update "active" state to "activated"
 - Add test for onactivate event

Honry committed:
Make tests clear

Honry committed:
Remove duplicated tests

dontcallmedom commented:
Merging as discussed on DAS call

Honry committed:
Fix compatibility issue between idlharness.js and chromium
